Added methods for nilpotent, abelian and cyclic numbers.

OrestisVaggelis opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Implemented methods to check if a number is nilpotent, abelian or cyclic. Details about the algorithm can be found in
[this](http://www2.math.ou.edu/~shankar/papers/nil2.pdf) paper.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* combinatorics
  * Added methods for nilpotent, abelian and cyclic numbers.
<!-- END RELEASE NOTES -->

OrestisVaggelis committed:
groups: added methods for nilpotent, abelian and cyclic numbers

Added methods to check if a positive integer 'n' is nilpotent,
abelian or cyclic. A number 'n' is said to be nilpotent
(abelian/cyclic) if every group of order 'n' is nilpotent
(abelian/cyclic). The algorithm is presented in the paper
"Nilpotent Numbers" by Pakianathan J. and  Shankar K.

OrestisVaggelis committed:
removed trailing whitespaces

OrestisVaggelis committed:
updated author details in .mailmap

OrestisVaggelis committed:
updated author details in .mailmap

OrestisVaggelis committed:
Update .mailmap/AUTHORS

OrestisVaggelis committed:
Update .mailmap/AUTHORS

oscarbenjamin commented on AUTHORS:
```diff
@@ -1133,3 +1133,13 @@ Tom Fryers <61272761+TomFryers@users.noreply.github.com>
+Luis Talavera <luisfertalavera15@gmail.com>
+Arie Bovenberg <a.c.bovenberg@gmail.com>
+Carson McManus <carson.mcmanus1@gmail.com>
+Jack Schmidt <1107865+jackschmidt@users.noreply.github.com>
```
The changes to this file should be reverted. The AUTHORS file is updated at the time of the next release. It's only necessary to add your details in the .mailmap file.

oscarbenjamin commented on .mailmap:
```diff
@@ -831,7 +831,7 @@ Ondřej Čertík <ondrej@certik.cz> <ondrej.certik@gmail.com>
 Ondřej Čertík <ondrej@certik.cz> ondrej.certik <devnull@localhost>
 Or Dvory <gidesa@gmail.com>
-OrestisVaggelis <orestis22000@gmail.com>
+OrestisVaggelis <orestis22000@gmail.com> Orestis Vaggelis <orestisvaggelis@mail.com>
```
Which name and email do you want to appear in the AUTHORS file?

OrestisVaggelis replied:
I'd prefer Orestis Vaggelis <orestisvaggelis@mail.com> to appear, but without the first one, I was getting an error during the tests.

oscarbenjamin replied:
Then the above should be reversed like:
```
Orestis Vaggelis <orestisvaggelis@mail.com>  OrestisVaggelis <orestis22000@gmail.com>
```
Also you should remove the line `OrestisVaggelis <orestis22000@gmail.com>` from the AUTHORS file. The new name will then be added to AUTHORS automatically at the time of the next release.

jksuom commented on sympy/combinatorics/group_numbers.py:
```diff
@@ -0,0 +1,119 @@
+    ==========
+
+    .. [1] Pakianathan, J., Shankar, K., Nilpotent Numbers,
+           The American Mathematical Monthly
```
There should be more details like year, issue, pages..

OrestisVaggelis committed:
updated .mailmap and AUTHORS files and added more details for the reference

oscarbenjamin commented on AUTHORS:
```diff
@@ -1133,3 +1132,9 @@ Tom Fryers <61272761+TomFryers@users.noreply.github.com>
 Zouhair <zouhair.mahboubi@gmail.com>
 zzj <29055749+zjzh@users.noreply.github.com>
 shubhayu09 <guptashubhayu601@gmail.com>
+Siddhant Jain <siddhantashoknagar@gmail.com>
+Tirthankar Mazumder <63574588+wermos@users.noreply.github.com>
+Sumit Kumar <mr.sumitkrr@gmail.com>
+Shivam Sagar <technoshivam12@gmail.com>
+Gaurav Jain <gjain369@gmail.com>
+Andrii Oriekhov <andriyorehov@gmail.com>
```
These lines should also be removed.

OrestisVaggelis committed:
reverted changes made in the AUTHORS file

OrestisVaggelis commented:
Thank you for your feedback. I've now reverted the changes made in the AUTHORS file, updated the .mailmap file and added more information about the reference.

oscarbenjamin commented on AUTHORS:
```diff
@@ -4,7 +4,7 @@ those who explicitly didn't want to be mentioned. People with a * next
 file is generated automatically by running `./bin/authors_update.py`.
 
-There are a total of 1127 authors.
+There are a total of 1137 authors.
```
This change also needs to be reverted.

When looking at the PR you can click "files changed" at the top to see what is actually being changed.

OrestisVaggelis replied:
So there should be 1126 authors now, since I removed myself, right?

oscarbenjamin replied:
The number will be updated at the time of the next release. For now it is best just to leave it unchanged.

The reason for not changing things like this is that if any two PRs changes the same line of the same file then there will be merge conflicts. If every PR tries to change this line then every PR will have merge conflicts.

OrestisVaggelis committed:
updated AUTHORS file

oscargus commented:
Should this be Abelian group in the docs? Wolfram seems to use Abelian, while Wikipedia and many other sources use abelian. Since Abel is a person it is not obvious to me as non-native.

OrestisVaggelis commented:
> Should this be Abelian group in the docs? Wolfram seems to use Abelian, while Wikipedia and many other sources use abelian. Since Abel is a person it is not obvious to me as non-native.

Other docs in sympy seem to use abelian as well (see permutation groups for example). Also the paper referenced uses a lower case a too.
